Dalek refactoring -- EXTERMINATE ALL THE CODE!

Rembane opened the pull request:
I did quite some refactoring to get to grips with how Rust works and this is the result.

I haven't added or removed any functionality (I hope). The idea throughout the refactoring has been to make the codebase leaner but to keep the functionality the same.

I'd love to get feedback on the code or just your thoughts on this endeavour. :)

Rembane committed:
Refactored the Rust Snakebot codebase.

Normal logging now logs to stdout.
Quite handy to get an overview of what is happening.

Renamed structs and constructors and added a constructor.

  * Renamed Inbound::GameLinkEvent to Inbound::GameLink.
  * Renamed the struct GameResultSnake to GameResult.
  * Added Inbound::GameResult.

Rewrote the message handling to become more succinct.
Sum types really excel at modeling this problem.

Turned the default_gamesettings function into a Default impl.

Made the snake smaller and prettier.

  * Replaced some direction checking code with less code.
  * Added logging messages to all callbacks.

Added as_movement_delta to the Direction impl.

Added nice error messages to the message parsing.

Moved some code in maputil to increase the DRY-factor.

mbark commented:
Very nice with some pull requests! 👏 Sorry for not getting back to you earlier (haven't checked my computer party due to easter) but I'll do so this week!

It would be great if you could do a breakdown of some of the changes you have done and the reason behind them considering the PR is decently large (`+142 −261`).

Rembane commented:
No worries. :)

The 30k feet overview is: I have removed as much code as possible while preserving the functionality.

And here's a description of why I have done what I have done in each commit: 

- 6d31312 -- direction_as_movement_delta fits really nicely into the Direction enum, so I moved it there and removed some ampersands. 
- 677d01c -- Having logging in stdout makes testing the snake straightforward since you can just use `cargo run` and then look at the log in realtime in the same terminal as you ran the command.
- c8e4fca -- I renamed `GameLinkEvent` to `GameLink` since it says the same thing with fewer characters. I also added `GameResult` since it was an event that happened now and then but hadn't been added.
- 2ebbce2 -- This is one of the biggest commits. The thought behind the changes here is to remove as much code as possible for message handling while still preserving the functionality. I did this by creating an `Outbound` enum, put all the outbound messages in it, made a succinct deserializing helper function called `handle_inbound_message` for `Inbound` messages and a function to serialize `Outbound` messages called `render_outbound_message`. Then I removed all functions used to handle a single message. Almost all string constants were also moved into match blocks in the functions that use them. 
- ea90428 -- I saw that I could use the `Default` trait for `GameSettings` and this felt so elegant that I just had to do it.
- f98231f -- I removed code from the snake to make it easier to understand and work with.
- 54439ab -- Just prettier error messages when parsing messages. I still think it's valuable for the program to panic if it gets something unfamiliar though.
- 21d5dbb -- I made the direction serializer less repetitive to make it easier to read and reason about.

That's all! If you have some feedback or questions, just ask! ^^

mbark commented on src/messages.rs:
```diff
@@ -44,97 +15,60 @@ pub enum Inbound {
+    let m            = v.as_object_mut().expect("I expected an object. Got something else.");
+    let typ          = m.remove("type").expect("This object doesn't contain the key `type`.")
+                        .as_str().expect("I expected a string. That was not a string.")
+                        .rsplit('.').next().expect("I expected a string containing at least one dot. This string didn't.")
```
I think a better wording here is one that indicates what went wrong, i.e.:
`Unexpected format for the message's type, expected at least one dot. Value was: {:?}`. The same goes for all of the strings in the expect here, they should be reworded to avoid using `I` and instead give an illustrative error for when they eventually get thrown in the user's face.

Otherwise, as I said... God, I wish I thought of this earlier. I spent so many hours making JSON parsing work...

Rembane replied:
Good point, I fix!

mbark commented on src/messages.rs:
```diff
@@ -44,97 +15,60 @@ pub enum Inbound {
+                        .replace("Event", "");
+    let mut m2 = Map::new();
+    m2.insert(typ, Value::Object(m.clone()));
+    from_value(Value::Object(m2))
```
Could you explain what is going on here? (Is there no easier way of doing this? If not, I think another comment is warranted)

Rembane replied:
There might be. `from_iter` is a good candidate. Let me get back to you on this one.

Rembane replied:
It was, and there was a really succinct way of expressing this, albeit bracket heavy.

mbark commented on src/snake.rs:
```diff
@@ -47,14 +38,14 @@ impl Snake {
 
     pub fn on_invalid_playername(&self, _: &InvalidPlayerName) {
-
+        debug!(target: LOG_TARGET, "I'm sorry, I cannot let you do that Dave.");
```
While these are clever I think it is generally better to stick to boring but informative defaults for these, i.e.:
`{:?} is not a valid player name!`

Rembane replied:
Since this is in a part of the program the end user should tinker with, I think it's alright to have corny error messages, but I will change it to something reasonable instead. :)

mbark reviewed (commented):
First of all: these are very nice improvements 👊! How did you find it working with the project in general? Since it's only been me in the code before I was afraid it would be a bit... illogical at places.

I will also say that the change where you pick the type from the JSON and then remove it... Why did I not think of that? That would have saved me several hours of headache trying to parse JSON with reserved keywords.

I have written some specific comments where I thought it appropriate but I also have one general comment:
- In some places you single-letter variable names like `m` or `d`, in general I have tried to avoid these and instead let things have the proper names they deserve (i.e. `message` (or `msg` if you want to) and `direction`).

Overall though as I said, it looks great! You didn't by chance try using this client to write an actual snake bot? (If you did, how did you find it was using the API? And specifically the `maputil` API?)

mbark commented:
Go through my comments when you have time and update (or don't) the code as you see fit and then squash it into one commit and I think we should merge this!

Rembane commented on src/maputil.rs:
```diff
@@ -11,31 +12,35 @@ pub enum Tile<'a> {
-        &Direction::Left => (-1, 0),
-        &Direction::Right => (1, 0)
+impl Direction {
+    pub fn direction_as_movement_delta(&self) -> (i32,i32) {
```
I see your point, but I think the ergonomics of `my_direction.direction_as_movement_delta()` are excellent. And I agree to 100% that `as_movement_delta()`is a better name.

Rembane commented:
> First of all: these are very nice improvements :fist_oncoming:! How did you find it working with the project in general? Since it's only been me in the code before I was afraid it would be a bit... illogical at places.

The code had a distinct C++/Java feeling to it, that is, verbose but otherwise fine. It was a really good learning experience to get to know Rust and this project at the same time. I had fun and enjoyed it!

> I will also say that the change where you pick the type from the JSON and then remove it... Why did I not think of that? That would have saved me several hours of headache trying to parse JSON with reserved keywords.

^___^ the next time you will. 

I stole the idea from the prototype of snake bot I built in Haskell. An enum type is a good fit and then you "just" have to find which constructor to use. Encode all the things in the type system! :D

> I have written some specific comments where I thought it appropriate but I also have one general comment:
> 
> In some places you single-letter variable names like m or d, in general I have tried to avoid these and instead let things have the proper names they deserve (i.e. message (or msg if you want to) and direction).

It's a Spartan programming/Haskell habit I have. I just can't be bothered with descriptive names for values with a really small scope. But I'll change them to make the code easier to understand.

> Overall though as I said, it looks great! You didn't by chance try using this client to write an actual snake bot? (If you did, how did you find it was using the API? And specifically the maputil API?)

I didn't know enough Rust to actually participate in the competition, I grabbed a piece of pie, watched other snakes battle it out and then went home and refactored the Rust client. :D

Rembane commented on src/messages.rs:
```diff
@@ -44,97 +15,60 @@ pub enum Inbound {
+    let typ          = m.remove("type").expect("This object doesn't contain the key `type`.")
+                        .as_str().expect("I expected a string. That was not a string.")
+                        .rsplit('.').next().expect("I expected a string containing at least one dot. This string didn't.")
+                        .replace("Event", "");
```
Good point! I'll document this!

Rembane commented:
It is done.

Rembane commented:
I was a bit too tired this morning when I pushed the last commit, so it had some errors which I fixed now.

mbark commented:
Merged! Good job @Rembane 👏👏👏

Rembane commented:
@mbark No worries. ^^
